src/coreComponents/finiteVolume/CellElementStencilTPFA.hpp

```diff
@@ -215,84 +272,204 @@ GEOS_HOST_DEVICE
   for( localIndex ke = 0; ke < 2; ++ke )
   {
     weight[0][ke] = m_transMultiplier[iconn] * value * (ke == 0 ? 1 : -1);
+    dWeight_dVar[0][ke] = 0;
+  }
+}
+
+GEOS_HOST_DEVICE
+inline void
+CellElementStencilTPFAWrapper::computeWeightsBase( localIndex const iconn,
+                                                   localIndex const icell,
+                                                   real64 & halfWeight,
+                                                   arraySlice3d< real64 const > const & coefficient ) const
+{
+
+  localIndex const ei = m_elementIndices[iconn][icell];
+
+  halfWeight = m_weights[iconn][icell];
 
-    real64 const dValue_dVar = meanPermCoeff * dHarmonicWeight_dVar[ke] + (1 - meanPermCoeff) * dArithmeticWeight_dVar[ke];
-    dWeight_dVar[0][ke] = m_transMultiplier[iconn] * dValue_dVar;
+  // Proper computation
+  real64 faceNormal[3], cellToFaceVec[3];
+  // previously was normalized in container
+  LvArray::tensorOps::copy< 3 >( cellToFaceVec, m_cellToFaceVec[iconn][icell] );
+  LvArray::tensorOps::normalize< 3 >( cellToFaceVec );
+
+  LvArray::tensorOps::copy< 3 >( faceNormal, m_faceNormal[iconn] );
+  if( LvArray::tensorOps::AiBi< 3 >( cellToFaceVec, faceNormal ) < 0.0 )
+  {
+    LvArray::tensorOps::scale< 3 >( faceNormal, -1 );
   }
+
+  real64 faceConormal[3];
+  LvArray::tensorOps::hadamardProduct< 3 >( faceConormal, coefficient[ei][0], faceNormal );
+  halfWeight *= LvArray::tensorOps::AiBi< 3 >( cellToFaceVec, faceConormal );
+
+  // correct negative weight issue arising from non-K-orthogonal grids
+  if( halfWeight < 0.0 )
+  {
+    LvArray::tensorOps::hadamardProduct< 3 >( faceConormal,
+                                              coefficient[ei][0],
+                                              cellToFaceVec );
+    halfWeight = m_weights[iconn][icell];
+    halfWeight *= LvArray::tensorOps::AiBi< 3 >( cellToFaceVec, faceConormal );
+  }
+
+}
+
+GEOS_HOST_DEVICE
+inline void
+CellElementStencilTPFAWrapper::
+  computeWeights( localIndex const iconn,
+                  CoefficientAccessor< arrayView3d< real64 const > > const & coefficient,
+                  CoefficientAccessor< arrayView3d< real64 const > > const & dCoeff_dVar,
+                  real64 (& weight)[1][2],
+                  real64 (& dWeight_dVar )[1][2] ) const
+{
+  GEOS_UNUSED_VAR( dCoeff_dVar );
+
+  real64 halfWeight[2];
+
+  // real64 const tolerance = 1e-30 * lengthTolerance; // TODO: choice of constant based on physics?
+
+  for( localIndex i = 0; i < 2; ++i )
+  {
+    localIndex const er = m_elementRegionIndices[iconn][i];
+    localIndex const esr = m_elementSubRegionIndices[iconn][i];
+
+    computeWeightsBase( iconn, i, halfWeight[i], coefficient[er][esr] );
+
+  }
+
+  averageWeights( iconn, weight, dWeight_dVar, halfWeight );
+
 }
 
+GEOS_HOST_DEVICE
+inline void
+CellElementStencilTPFAWrapper::
+  computeVelocity( localIndex iconn,
```

Comment:
I don't love this function being defined in here. The stencil is also used for non advective fluxes (heat conduction for example) and this does not make much sense at that point. This function only uses the stencil to get the connectivity. I would make a free function that uses an instance of the object instead of a method in here.